pkg/scm/git/git.go

```diff
@@ -62,18 +84,272 @@ func stringInSlice(s string, slice []string) bool {
 // ValidCloneSpec determines if the given string reference points to a valid git
 // repository
 func (h *stiGit) ValidCloneSpec(source string) bool {
-	url, err := url.Parse(source)
-	if err != nil {
-		return false
+	details, _ := ParseFile(source)
+	if details.FileExists && !details.BadRef {
+		return true
+	}
+
+	return h.ValidCloneSpecRemoteOnly(source)
+}
+
+// ValidCloneSpecRemoteOnly determines if the given string reference points to a valid remote git
+// repository; valid local repository specs will result in a return of false
+func (h *stiGit) ValidCloneSpecRemoteOnly(source string) bool {
+	_, err := ParseSSH(source)
+	if err == nil {
+		return true
 	}
 
-	if stringInSlice(url.Scheme, allowedSchemes) {
+	url, err := ParseURL(source)
+	if url != nil && err == nil && url.Scheme != "file" {
 		return true
 	}
 
-	// support 'git@' ssh urls and local protocol without 'file://' scheme
-	return url.Scheme == "" && (strings.HasSuffix(source, ".git") ||
-		(strings.HasPrefix(source, "git@") && gitSshURLExp.MatchString(source)))
+	return false
+}
+
+// MungeNoProtocolURL will take a URL returned from net/url.Parse and make
+// corrections to the URL when no protocol is specified in the Scheme, where there
+// are valid protocol-less git url spec formats that result in either file:// or ssh:// protocol usage;
+// if an explicit protocol is already specified, then the
+// URL is left unchaged and the method simply returns with no error reported,
+// since the URL is
+func (h *stiGit) MungeNoProtocolURL(source string, uri *url.URL) error {
+	if uri == nil {
+		return nil
+	}
+
+	// only deal with protocol-less url's
+	if uri.Scheme != "" {
+		return nil
+	}
+
+	details, mods := ParseFile(source)
+
+	if details.BadRef {
+		return errors.New(fmt.Sprintf("bad reference following # in %s", source))
+	}
+	if !details.FileExists {
+		mods2, err := ParseSSH(source)
+		if err != nil {
+			glog.Errorf("ssh git clone spec error: %v", err)
+			return err
+		}
+		mods = mods2
+	}
+
+	// update the either file or ssh url accordingly
+	if mods != nil {
+		if len(mods.user) > 0 {
+			uri.User = url.User(mods.user)
+		}
+		if len(mods.scheme) > 0 {
+			uri.Scheme = mods.scheme
+		}
+		if len(mods.host) > 0 {
+			uri.Host = mods.host
+		}
+		if len(mods.path) > 0 {
+			uri.Path = mods.path
+		}
+		if len(mods.ref) > 0 {
+			uri.Fragment = mods.ref
+		}
+	}
+	return nil
+}
+
+// ParseURL sees if the spec is properly handled by golang's URL parser,
+// with a valid, explicit protocol specified,
+// and deals with some of the known issues we have seen with the logic
+func ParseURL(source string) (*url.URL, error) {
+	uri, err := url.Parse(source)
+	if err != nil {
+		return nil, err
+	}
+
+	// to work around the url.Parse bug, deal with the fact it seemed to split on just :
+	if strings.Index(source, "://") == -1 && uri.Scheme != "" {
+		uri.Scheme = ""
+		return nil, nil
```

Comment:
should this not return an error also?  (that it's not a valid url with a protocol)